backend/src/index.ts

```diff
@@ -106,12 +106,21 @@ const processSourceBlocks = (target: Target) => async (source: Source) => {
       archives.forEach(async archive => {
         let lastBlockProcessingReportAt = Date.now();
         let processedBlocks = 0;
-
+        let pendingTxs = [];
         let nonce = (await target.api.rpc.system.accountNextIndex((archive.signer as KeyringPair).address)).toBn();
+
         for await (const blockData of archive.getBlocks()) {
-          target.sendBlockTx(blockData, nonce);
+          const tx = target.sendBlockTx({ ...blockData, nonce });
+
+          pendingTxs.push(tx);
+
           nonce = nonce.add(new BN(1));
 
+          if (pendingTxs.length > 500) {
+            await Promise.all(pendingTxs);
```

Comment:
Here is the challenge: you are waiting for a batch of 500 transactions, while ideally you would instead allow up to 500 transactions to be **running concurrently** instead. So you don't need to wait for anything if the performance is above 500 tx/s.

So you need to be able to know how many concurrent txs there are (which is easy to do) and if the number reaches a limit, in this case 500, you will wait for the oldest transaction to finish before sending a new one. You don't need to wait for all of them because in that case you will waste valuable time during which you could do something useful like reading and preparing the next transaction to be sent.